### ISSUE TITLE
FEATURE REQUEST - 2FA to open application

### ISSUE THREAD

**Issue by pir8radio** [mRemoteNG/mRemoteNG#2325]
Encryption of the connections file is great, but I would feel even safer if I could enable 2fa to even open the application,  I don't think it needs a user/pass to login, just a function to enable something like using google authenticator or Authy,  to 2fa into the app..  I do not save login info within mRemoteNG today because if i loose my laptop, its pretty easy to get into windows.    then they have the key to my servers.

**Comment by Kvarkas** [comment]
that is palned for user profiles

**Comment by My1** [comment]
the biggest issue tho is this needs to be made in a way that isnt pure snake oil.

TOTP (as in the algo that Google Authenticator etc generally use) only works because it is used to authenticate at remote systems which can deny giving you any data in the first place.

However with mRem your keys/passwords etc, if stored in there, are local, this means the only way to properly secure them is encryption. an Authentication-only approach would not really work as an attacker could just bypass the application and get to your config files directly and decrypt them with the password if it is encrypted and they have the password already.

After All as mRem cannot know when you try to decrypt, a code generated based on the current time, wouldnt really help with anything.

Keepass has an extension that abuses HOTP in an way that is both crazy and genius, HOTP is basically the parent of TOTP which uses the same base algo but uses an incrementing counter instead of the time, so the application can predict the next TOTP and use that for encryption, although iirc the plugin uses a sequence of a few instead of just a single one, as just a million possibilities is a joke.

https://keepass.info/plugins.html#otpkeyprov

other methods that could work are
1) smartcard based stuff (doesnt even have to be an actual card, there are several USB-Sticks that are smartcards inside like Nitrokeys or Yubikeys
2) Yubikey-HMAC (although the Yubi generally has a smartcard as well so the smartcard approach is more generic)
3) FIDO2 -> HMAC-Secret (likely not easy to use as windows 10 blocks direct access to FIDO devices and their Win-Hello based FIDO API doesnt iirc pass through the hmac-secret extension)
4) WinHello (PIN/Biometric) based stuff, which with a TPM can add a forced retry counter, as well as likely no easy way to try out PINs outside the normal path. biggest issue, the TPM is comparatively easy to nuke during more or less normal use, like a Bios Update iirc is generally likely to knock out its data.
5) (not really a good option but for the sake of completeness (and laughter)) some completely overengineered app-based solution that uses the phone and sends decryption keys over half the internet through THE CLOUD.

**Comment by pir8radio** [comment]
> the biggest issue tho is this needs to be made in a way that isnt pure snake oil.
> 
> TOTP (as in the algo that Google Authenticator etc generally use) only works because it is used to authenticate at remote systems which can deny giving you any data in the first place.
> 
> However with mRem your keys/passwords etc, if stored in there, are local, this means the only way to properly secure them is encryption. an Authentication-only approach would not really work as an attacker could just bypass the application and get to your config files directly and decrypt them with the password if it is encrypted and they have the password already.
> 
> After All as mRem cannot know when you try to decrypt, a code generated based on the current time, wouldnt really help with anything.
> 
> Keepass has an extension that abuses HOTP in an way that is both crazy and genius, HOTP is basically the parent of TOTP which uses the same base algo but uses an incrementing counter instead of the time, so the application can predict the next TOTP and use that for encryption, although iirc the plugin uses a sequence of a few instead of just a single one, as just a million possibilities is a joke.
> 
> https://keepass.info/plugins.html#otpkeyprov
> 
> other methods that could work are
> 
> 1. smartcard based stuff (doesnt even have to be an actual card, there are several USB-Sticks that are smartcards inside like Nitrokeys or Yubikeys
> 2. Yubikey-HMAC (although the Yubi generally has a smartcard as well so the smartcard approach is more generic)
> 3. FIDO2 -> HMAC-Secret (likely not easy to use as windows 10 blocks direct access to FIDO devices and their Win-Hello based FIDO API doesnt iirc pass through the hmac-secret extension)
> 4. WinHello (PIN/Biometric) based stuff, which with a TPM can add a forced retry counter, as well as likely no easy way to try out PINs outside the normal path. biggest issue, the TPM is comparatively easy to nuke during more or less normal use, like a Bios Update iirc is generally likely to knock out its data.
> 5. (not really a good option but for the sake of completeness (and laughter)) some completely overengineered app-based solution that uses the phone and sends decryption keys over half the internet through THE CLOUD.

the store can already be encrypted right?   I just want a simple 2FA to decrypt that file.  For google authenticator and Authy no internet is needed let alone any encryption keys sent over the net.  Its based on time and shared keys.
![image](https://user-images.githubusercontent.com/16748696/214973626-5bd04777-8ee6-4373-92a3-44a1da0e48cb.png)

**Comment by simonai1254** [comment]
> the store can already be encrypted right? I just want a simple 2FA to decrypt that file. For google authenticator and Authy no internet is needed let alone any encryption keys sent over the net. Its based on time and shared keys.

It is correct that the TOTP protocol does not require any Internet. The problem is (like @My1 statet) that you are not connecting to a backend that validates the second factor, but you are on that backend. This means an attacker that has access to your system and therefore the connection file can run the attack by writing a decryptor that just ignores the 2FA part and use that. 
The 2FA implemented like you would like to have (TOTP) only serves as deterrence, there is no technical security benefit. 
The only way to properly implement this is using hardware 2FA, like the options proposed by @My1. Additionally one might look at a second key file stored on a USB drive as a second factor (like KeePass offers), but a proper certificate solution with hardware protection would be best.

**Comment by My1** [comment]
exactly. the encryption of that file can only use your password, or some otherwise static key to actually encrypt that file, as you cannot have that thing change until you go and decrypt it again.

things like TOTP I generally call a "data valve" you can use it to deny access but that really only works when the attacker is on the other side of that valve.

in Fact I have written a big thing about this entire thing when a kinda reputable thing like Steganos went and implemented TOTP
https://blog.my1.dev/steganos-privacy-suite-19-is-a-joke

**Comment by Kvarkas** [comment]
My idea is separate such in profiles so we remove responsibility for sensitive data from mremote (even if we try to make it as much secure as posible) one way is db - then users could have they own policy on that, also if it still connection file it may be stored on user premises, and in additional passwords could be stored somewhere esle. So logical link will be looks as that:
App -> User (local login, or network login) + 2FA -> User profile (where is always default user profile if user not able to authorise on any else) -> Connection sources (file - local or remote, db - local or remote) could be several joined and devided by ACL -> Connection profile (local or remote) -> Connection data (local or remote) So potential user login into windows and that will be his default login, on default user profile, with empty connection profile. User may add (create, add, edit depending on acl) connection source, or select assigned (if any) and add (create, add, edit) connection profile where he state that for item dev server connection detais as follows, password as follows (from source X) default protocols and settings as follows. Add next - uat server - use created profile below (select, assign) etc. If login by another user - it have no access to that info as it was encoded and e ncrypted by another user. dead end. However if user creating connection profile set it as shareble, it will be visble to user2, but he could use it depending on acl - us as it is (without dispousing connection details and credentilas) able to amend, able to delete, able to add. etc. So its a bit big chunk of work what will be happening but it should make all users happy - alow simple usage for home users and give flexibility for corporate users (they could add they own policy on what user could and what not) so far long story short :D

**Comment by My1** [comment]
okay if the data is actually brought in from remote rather than stored locally, auth-only 2FA would actually work

**Comment by devNull2and1** [comment]
hello, didn't see that open issue and started discussion for feature request:

https://github.com/orgs/mRemoteNG/discussions/2572#discussioncomment-8818392

would be nice when this feature will be added :O)

**Comment by robertpopa22** [comment]
**Fix available for testing**

**Commit:** `dd5f4dc7` (fork `robertpopa22/mRemoteNG`, branch `main`)
**What changed:** FEATURE REQUEST - 2FA to open application

A beta build (v1.81.0-beta.4) including this fix is available from the fork's Releases page.

---

> *We sincerely appreciate the mRemoteNG maintainers and the wonderful community that has kept this project alive and thriving for over a decade. Your work powers thousands of IT professionals worldwide, including us. Our fork builds entirely on your foundation — we're preparing a consolidated Pull Request to contribute these improvements back to the official project. Thank you for everything you've built.*
>
> *— robertpopa22, community contributor*
